src/node/ledger_secrets.h

```diff
@@ -3,118 +3,283 @@
+    LedgerSecret() = default;
+
+    LedgerSecret(const LedgerSecret& other) :
+      raw_key(other.raw_key),
+      key(std::make_shared<crypto::KeyAesGcm>(other.raw_key))
+    {}
+
+    LedgerSecret(std::vector<uint8_t>&& raw_key_) :
+      raw_key(raw_key_),
+      key(std::make_shared<crypto::KeyAesGcm>(std::move(raw_key_)))
+    {}
+  };
+
+  inline LedgerSecret make_ledger_secret()
+  {
+    return LedgerSecret(tls::create_entropy()->random(crypto::GCM_SIZE_KEY));
+  }
+
+  using VersionedLedgerSecret = std::pair<kv::Version, LedgerSecret>;
+  using LedgerSecretsMap = std::map<kv::Version, LedgerSecret>;
+
+  struct VersionedLedgerSecrets
+  {
+    LedgerSecretsMap secrets;
+
+    VersionedLedgerSecrets() = default;
+
+    VersionedLedgerSecrets(LedgerSecretsMap&& secrets_) :
+      secrets(std::move(secrets_))
+    {}
+
+    bool operator==(const VersionedLedgerSecrets& other) const
     {
-      master = tls::create_entropy()->random(MASTER_KEY_SIZE);
+      return secrets == other.secrets;
     }
-
-    LedgerSecret(const std::vector<uint8_t>& master_) : master(master_) {}
   };
 
   class LedgerSecrets
   {
-  public:
-    struct VersionedLedgerSecret
+  private:
+    Secrets& secrets_table;
+
+    std::optional<NodeId> self = std::nullopt;
+
+    SpinLock lock;
+    VersionedLedgerSecrets ledger_secrets;
+    std::optional<LedgerSecretsMap::iterator> commit_key_it = std::nullopt;
+
+    LedgerSecretsMap::iterator get_encryption_key_it(kv::Version version)
     {
-      kv::Version version;
-      LedgerSecret secret;
+      // Encryption key for a given version is the one with the highest version
+      // that is lower than the given version (e.g. if encryption_keys contains
+      // two keys for version 0 and 10 then the key associated with version 0
+      // is used for version [0..9] and version 10 for versions 10+)
 
-      bool operator==(const VersionedLedgerSecret& other) const
+      auto search_begin =
+        commit_key_it.value_or(ledger_secrets.secrets.begin());
+
+      auto search = std::upper_bound(
+        search_begin,
+        ledger_secrets.secrets.end(),
+        version,
+        [](auto a, const auto& b) { return b.first > a; });
+      if (search == search_begin)
+      {
+        throw std::logic_error(fmt::format(
+          "kv::TxEncryptor: could not find ledger encryption key for seqno {}",
+          version));
+      }
+      return --search;
+    }
+
+    void take_dependency_on_secrets(kv::Tx& tx)
+    {
+      // Ledger secrets are not stored in the KV. Instead, they are
+      // cached in a unique LedgerSecrets instance that can be accessed without
+      // reading the KV. However, it is possible that the ledger secrets are
+      // updated (e.g. rekey tx) concurrently to their access by another tx. To
+      // prevent conflicts, accessing the ledger secrets require access to a tx
+      // object, which must take a dependency on the secrets table.
+      auto v = tx.get_view(secrets_table);
```

Comment:
```suggestion
      auto v = tx.get_view<Secrets>(Tables::SECRETS);
```

If we lookup this table by name, we can remove the ref member, and remove the constructor argument, and remove some of the changes to test cases where this no longer needs to be passed.

Reply:
Ah! True. I forgot we could do that 👍